cli/src/command/finalize.rs

```diff
@@ -63,49 +64,73 @@ pub struct Finalize {
+        let mut public_configs = BTreeMap::new();
         let first = config_iter
             .next()
-            .ok_or_else(|| format_err!("Expected at least one config in PUBLIC_CONFIGS"))?;
-        let common = first.common;
-        map.insert(first.node.validator_keys.consensus_key, first.node);
+            .ok_or_else(|| format_err!("Expected at least one config in <public-configs>"))?;
+        let consensus_key = Self::get_consensus_key(&first)?;
+        public_configs.insert(consensus_key, first.clone());
 
         for config in config_iter {
-            if common != config.common {
-                bail!("Found config with different common part.");
+            if first.consensus != config.consensus {
+                bail!(
+                    "Found public configs with different consensus configuration.\
+                     Make sure the same template config was used for generation.\
+                     {:#?} \nnot equal to\n {:#?}",
+                    first.consensus,
+                    config.consensus
+                );
+            }
+            if first.general != config.general {
+                bail!(
+                    "Found public configs with different general configuration.\
+                     Make sure the same template config was used for generation.\
+                     {:#?} \nnot equal to\n {:#?}",
+                    first.general,
+                    config.general
+                );
             };
-            if map
-                .insert(config.node.validator_keys.consensus_key, config.node)
-                .is_some()
-            {
-                bail!("Found duplicate consensus keys in PUBLIC_CONFIGS");
+
+            let consensus_key = Self::get_consensus_key(&config)?;
+            if public_configs.insert(consensus_key, config).is_some() {
+                bail!(
+                    "Found duplicated consensus keys in <public-configs>: {:?}",
+                    consensus_key
+                );
             }
         }
         Ok(ValidatedConfigs {
-            common,
-            public_configs: map.values().cloned().collect(),
+            common: first,
+            public_configs: public_configs.values().cloned().collect(),
         })
     }
 
+    fn get_consensus_key(config: &NodePublicConfig) -> Result<PublicKey, failure::Error> {
+        Ok(config
+            .validator_keys
+            .ok_or_else(|| format_err!("Expected validator keys in public config: {:#?}", config))?
+            .consensus_key)
+    }
+
     fn create_connect_list_config(
         public_configs: &[NodePublicConfig],
         secret_config: &NodePrivateConfig,
     ) -> ConnectListConfig {
         let peers = public_configs
             .iter()
             .filter(|config| {
-                config.validator_keys.consensus_key != secret_config.keys.consensus_pk()
+                config.validator_keys.unwrap().consensus_key != secret_config.keys.consensus_pk()
```

Comment:
Does it make sense to replace `unwrap()` with `except("Couldn't get validators keys")` ?